File: tooling/scripts/audit_ui.js
() => {
  const F = [];
  const add = (kind, detail) => F.push({ kind, detail });
  const name = (el) => {
    const c = (el.className && el.className.toString ? el.className.toString() : "").trim();
    return (el.tagName.toLowerCase() + (c ? "." + c.split(/\s+/).slice(0,2).join(".") : "")).slice(0, 44);
  };
  const scrollers = [...document.querySelectorAll("*")].filter((el) => {
    const s = getComputedStyle(el);
    const oy = s.overflowY, ox = s.overflowX;
    return ((oy === "auto" || oy === "scroll") && el.scrollHeight > el.clientHeight + 1)
        || ((ox === "auto" || ox === "scroll") && el.scrollWidth  > el.clientWidth  + 1);
  });

  // 1. Dead scroll space: room to scroll past the last thing in the box.
  for (const el of scrollers) {
    // Deepest painted bottom anywhere inside, not just the direct children's
    // boxes: a child that is shorter than its own content would otherwise make
    // real content below the fold look like empty space.
    const all = [...el.querySelectorAll("*")].filter((k) => {
      const r = k.getBoundingClientRect();
      const cs = getComputedStyle(k);
      return r.height > 0 && cs.visibility !== "hidden" && cs.display !== "none";
    });
    if (!all.length) continue;
    const host = el.getBoundingClientRect();
    const lastBottom = Math.max(...all.map((k) => k.getBoundingClientRect().bottom));
    const remaining = el.scrollHeight - el.clientHeight - el.scrollTop;   // still scrollable
    const belowFold = lastBottom - host.bottom;                          // content still to come
    const cs = getComputedStyle(el);
    // A scroller's own bottom padding is scrollable and is meant to be — reaching
    // the end of a document should show its bottom margin. Counting it made the
    // navigation rail report 12-24px of "dead" space on every short viewport,
    // which is the kind of standing false positive that gets a check ignored.
    const padBottom = parseFloat(cs.paddingBottom) || 0;
    const dead = Math.round(remaining - Math.max(0, belowFold) - padBottom);
    if (dead > 2) add("dead-scroll", `${name(el)} can scroll ${dead}px past its content`);
    if (cs.paddingTop !== "0px" || cs.paddingBottom !== "0px")
      add("scroller-padding", `${name(el)} pt=${cs.paddingTop} pb=${cs.paddingBottom} (counts as scrollable space)`);
  }

  // 2. Page-level overflow.
  const d = document.documentElement;
  if (d.scrollWidth  > d.clientWidth  + 1) add("page-overflow-x", `${d.scrollWidth - d.clientWidth}px`);
  if (d.scrollHeight > d.clientHeight + 1) add("page-overflow-y", `${d.scrollHeight - d.clientHeight}px`);

  // 3. Interactive elements clipped horizontally by their scroll parent.
  const xParent = (el) => {
    for (let p = el.parentElement; p; p = p.parentElement) {
      const o = getComputedStyle(p).overflowX;
      if (o === "auto" || o === "scroll" || o === "hidden") return p;
    }
    return null;
  };
  // Skip the drawer's contents only while it is genuinely off screen. The check is
  // narrow on purpose: written any looser it excused everything inside the rail,
  // and the phone drawer was never examined at all.
  const inClosedRail = (el) => {
    for (let p = el; p; p = p.parentElement) {
      if (!p.classList?.contains("rail")) continue;
      if (p.dataset.open === "true") return false;
      const cs = getComputedStyle(p);
      if (cs.position !== "fixed") return false;
      const r = p.getBoundingClientRect();
      return r.right <= 0 || r.left >= document.documentElement.clientWidth;
    }
    return false;
  };
  for (const el of document.querySelectorAll("button, a[href], input, [role='tab']")) {
    const b = el.getBoundingClientRect();
    if (b.width === 0 && b.height === 0) continue;
    if (inClosedRail(el)) continue;
    if (el.closest("[inert]")) continue;
    const host = xParent(el);
    const lim = host ? host.getBoundingClientRect() : { left: 0, right: d.clientWidth };
    if (b.right > lim.right + 1 || b.left < lim.left - 1)
      add("clipped-control", `${name(el)} x[${Math.round(b.left)},${Math.round(b.right)}] vs [${Math.round(lim.left)},${Math.round(lim.right)}]`);
  }

  // 4. Tap targets.
  for (const el of document.querySelectorAll("button, a[href], [role='tab'], input[type=checkbox]")) {
    const b = el.getBoundingClientRect();
    if (b.width === 0 || b.height === 0) continue;
    if (inClosedRail(el) || el.closest("[inert]")) continue;
    if (b.height < 24 || b.width < 24)
      add("tiny-target", `${name(el)} ${Math.round(b.width)}x${Math.round(b.height)}`);
  }

  // 5. Distorted images.
  for (const im of document.querySelectorAll("img")) {
    const b = im.getBoundingClientRect();
    if (!im.naturalWidth || !b.width || !b.height) continue;
    const fit = getComputedStyle(im).objectFit;
    if (fit !== "fill") continue;
    const want = im.naturalWidth / im.naturalHeight, got = b.width / b.height;
    if (Math.abs(want - got) / want > 0.02)
      add("stretched-image", `${im.getAttribute("src")?.slice(-28)} ${Math.round(b.width)}x${Math.round(b.height)} vs natural ${im.naturalWidth}x${im.naturalHeight}`);
  }

  // 6. A wide rail with no labels — the fault a clean run missed entirely.
  const rail = document.querySelector(".rail");
  if (rail && !inClosedRail(rail)) {
    const rw = rail.getBoundingClientRect().width;
    const rows = [...rail.querySelectorAll(".nav-row")];
    const hidden = rows.filter((row) => {
      const lab = row.querySelector(".nav-label");
      return lab && getComputedStyle(lab).display === "none";
    });
    if (rw > 120 && hidden.length === rows.length && rows.length > 0)
      add("rail-labels-hidden", `rail is ${Math.round(rw)}px wide with all ${rows.length} labels hidden`);
    for (const row of rows) {
      const b = row.getBoundingClientRect();
      const icon = row.querySelector(".nav-icon");
      if (!icon) continue;
      const ib = icon.getBoundingClientRect();
      // An icon parked at the far left of a wide empty row is the visual symptom.
      if (rw > 120 && b.width - (ib.right - b.left) > b.width * 0.6 && hidden.length)
        add("rail-row-empty", `${row.textContent.trim().slice(0,14) || "row"} ${Math.round(b.width)}px wide, icon only`);
    }
  }

  // 7. Hygiene.
  const ids = {};
  for (const el of document.querySelectorAll("[id]")) ids[el.id] = (ids[el.id] || 0) + 1;
  for (const [id, n] of Object.entries(ids)) if (n > 1) add("duplicate-id", `#${id} x${n}`);
  for (const im of document.querySelectorAll("img")) if (im.getAttribute("alt") === null) add("img-no-alt", name(im));
  for (const el of document.querySelectorAll("button")) {
    if (!el.textContent.trim() && !el.getAttribute("aria-label") && !el.getAttribute("title"))
      add("unlabelled-button", name(el));
  }

  // 8. Text clipped by its own box.
  for (const el of document.querySelectorAll("h1,h2,h3,p,span,button,a")) {
    if (el.children.length) continue;
    const s = getComputedStyle(el);
    if (s.overflow === "visible" && s.textOverflow !== "ellipsis") continue;
    if (el.scrollWidth > el.clientWidth + 1 && s.textOverflow !== "ellipsis")
      add("clipped-text", `${name(el)} "${el.textContent.trim().slice(0,24)}" ${el.scrollWidth}>${el.clientWidth}`);
  }

  return F;
}

File: tooling/scripts/audit_ui.py
"""Sweep the running app for layout faults, across viewports and screen states.

Written because a whole class of bug here is invisible to unit tests and easy to
miss by eye: a scroller that can be dragged past its content, a control clipped
outside its pane, an image stretched by a missing `height: auto`. Each one reads as
"the site feels buggy" without ever failing an assertion.

The check that earns its place is `dead-scroll` — scrollable distance beyond the
last painted pixel inside the box. It found a phone review screen whose entire
content column could be dragged 810px into empty background, because a stale
`overflow: visible` on the question list sent its overflow up to the first ancestor
that would take it.

Two lessons are baked into the metric itself. It measures the deepest painted
bottom among *all* descendants rather than the direct children's boxes, because a
child shorter than its own content otherwise makes real content below the fold look
like emptiness — which produced a 265px false positive on the upload screen. And it
reports padding on scrollers separately rather than counting it as a fault, because
a deliberate top inset and 810px of nothing are not the same finding.

Usage:
    python tooling/scripts/audit_ui.py <submission_id> [--base http://127.0.0.1:3001]

Needs a marked submission to audit the review screen, and the dev servers running.
Exits non-zero if anything but scroller padding is reported.
"""

import asyncio
from playwright.async_api import async_playwright
import argparse
from pathlib import Path

parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument("submission", help="A submission id that has been marked.")
parser.add_argument("--base", default="http://127.0.0.1:3001")
ARGS = parser.parse_args()
SUB = ARGS.submission
BASE = ARGS.base
PROBE = (Path(__file__).parent / "audit_ui.js").read_text()
SIZES = [("phone", 393, 852), ("phone-sm", 320, 568), ("tablet", 768, 1024),
         ("laptop", 1440, 900), ("laptop-short", 1440, 600), ("wide", 1920, 1080)]

async def audit(pg, label, extra=""):
    findings = await pg.evaluate(PROBE)
    if not findings:
        print(f"  {label}{extra}: clean")
    else:
        seen = {}
        for f in findings:
            seen.setdefault(f["kind"], []).append(f["detail"])
        print(f"  {label}{extra}:")
        for kind, details in seen.items():
            for det in details[:3]:
                print(f"     [{kind}] {det}")
            if len(details) > 3:
                print(f"     [{kind}] ... +{len(details)-3} more")
    return findings

async def main():
    all_f = {}
    async with async_playwright() as p:
        b = await p.chromium.launch()
        for label, w, h in SIZES:
            ctx = await b.new_context(viewport={"width": w, "height": h})
            pg = await ctx.new_page()

            print(f"\n=== {label} {w}x{h} — UPLOAD")
            await pg.goto(f"{BASE}/", wait_until="networkidle", timeout=60000)
            await pg.wait_for_timeout(900)
            all_f[f"{label}/upload"] = await audit(pg, "upload")

            print(f"=== {label} {w}x{h} — REVIEW")
            await pg.goto(f"{BASE}/review/{SUB}", wait_until="networkidle", timeout=60000)
            await pg.wait_for_timeout(2200)
            all_f[f"{label}/review-top"] = await audit(pg, "review", " (at top)")

            # Sheet tab on narrow, then scrolled to the very bottom.
            tab = pg.locator('.map-tab', has_text="Answer Sheet")
            if await tab.count() and await tab.first.is_visible():
                await tab.first.click(); await pg.wait_for_timeout(700)
            await pg.evaluate("()=>{const n=document.querySelector('.sheet-scroll'); if(n) n.scrollTop=n.scrollHeight;}")
            await pg.wait_for_timeout(700)
            all_f[f"{label}/review-bottom"] = await audit(pg, "review", " (sheet scrolled to end)")

            # Expand everything, which is where clipping tends to show up.
            ea = pg.locator('.q-head-action', has_text="Expand All")
            qt = pg.locator('.map-tab', has_text="Questions")
            if await qt.count() and await qt.first.is_visible():
                await qt.first.click(); await pg.wait_for_timeout(500)
            if await ea.count() and await ea.first.is_visible():
                await ea.first.click(); await pg.wait_for_timeout(800)
            all_f[f"{label}/review-expanded"] = await audit(pg, "review", " (all expanded)")

            # The navigation drawer, opened.
            #
            # This was the audit's blind spot: every check skipped elements inside a
            # closed rail, on the grounds that an off-canvas drawer is off-canvas by
            # design — and then nothing ever opened it. A 304px drawer containing
            # eight unlabelled icons sat there through a clean run.
            opener = pg.locator('button[aria-label="Open navigation"]')
            if await opener.count() and await opener.first.is_visible():
                await opener.first.click()
                await pg.wait_for_timeout(700)
                all_f[f"{label}/nav-drawer"] = await audit(pg, "nav drawer", " (open)")
            await ctx.close()
        await b.close()
    kinds = {}
    for v in all_f.values():
        for f in v:
            kinds[f["kind"]] = kinds.get(f["kind"], 0) + 1
    total = sum(kinds.values())
    print(f"\n===== {total} findings across {len(all_f)} states")
    for k, n in sorted(kinds.items(), key=lambda x: -x[1]):
        print(f"  {n:>4}  {k}")
    # Scroller padding is reported, not failed: a deliberate inset is not a bug.
    return 1 if any(k != "scroller-padding" for k in kinds) else 0


raise SystemExit(asyncio.run(main()))
